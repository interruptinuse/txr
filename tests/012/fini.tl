(load "../common")

(defstruct base nil
  others
  id
  (:static counter 0)
  (:postinit (me) (set me.id (inc me.counter)))
  (:fini (me)
    (put-line `@(typeof me):@{me.id} finalized`)
    [mapdo call-finalizers me.others]))

(defstruct derived base
  (:fini (me)
    (put-line `@(typeof me):@{me.id} derived fini`))
  (:postfini (me)
    (put-line `@(typeof me):@{me.id} derived postfini`)))

(unwind-protect
  (with-objects ((b (new base others (mapcar (ret (new derived)) (range 1 20)))))
    (put-line "inside with-objects"))
  (put-line "after with-objects"))

(mapcar (ret (new derived)) (range 1 20))
(sys:gc)
