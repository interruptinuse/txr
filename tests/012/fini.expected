inside with-objects
base:21 finalized
derived:1 derived fini
derived:1 finalized
derived:1 derived postfini
derived:2 derived fini
derived:2 finalized
derived:2 derived postfini
derived:3 derived fini
derived:3 finalized
derived:3 derived postfini
derived:4 derived fini
derived:4 finalized
derived:4 derived postfini
derived:5 derived fini
derived:5 finalized
derived:5 derived postfini
derived:6 derived fini
derived:6 finalized
derived:6 derived postfini
derived:7 derived fini
derived:7 finalized
derived:7 derived postfini
derived:8 derived fini
derived:8 finalized
derived:8 derived postfini
derived:9 derived fini
derived:9 finalized
derived:9 derived postfini
derived:10 derived fini
derived:10 finalized
derived:10 derived postfini
derived:11 derived fini
derived:11 finalized
derived:11 derived postfini
derived:12 derived fini
derived:12 finalized
derived:12 derived postfini
derived:13 derived fini
derived:13 finalized
derived:13 derived postfini
derived:14 derived fini
derived:14 finalized
derived:14 derived postfini
derived:15 derived fini
derived:15 finalized
derived:15 derived postfini
derived:16 derived fini
derived:16 finalized
derived:16 derived postfini
derived:17 derived fini
derived:17 finalized
derived:17 derived postfini
derived:18 derived fini
derived:18 finalized
derived:18 derived postfini
derived:19 derived fini
derived:19 finalized
derived:19 derived postfini
derived:20 derived fini
derived:20 finalized
derived:20 derived postfini
after with-objects
derived:41 derived fini
derived:41 finalized
derived:40 derived fini
derived:40 finalized
derived:39 derived fini
derived:39 finalized
derived:38 derived fini
derived:38 finalized
derived:37 derived fini
derived:37 finalized
derived:36 derived fini
derived:36 finalized
derived:35 derived fini
derived:35 finalized
derived:34 derived fini
derived:34 finalized
derived:33 derived fini
derived:33 finalized
derived:32 derived fini
derived:32 finalized
derived:31 derived fini
derived:31 finalized
derived:30 derived fini
derived:30 finalized
derived:29 derived fini
derived:29 finalized
derived:28 derived fini
derived:28 finalized
derived:27 derived fini
derived:27 finalized
derived:26 derived fini
derived:26 finalized
derived:25 derived fini
derived:25 finalized
derived:24 derived fini
derived:24 finalized
derived:23 derived fini
derived:23 finalized
derived:22 derived fini
derived:22 finalized
derived:22 derived postfini
derived:23 derived postfini
derived:24 derived postfini
derived:25 derived postfini
derived:26 derived postfini
derived:27 derived postfini
derived:28 derived postfini
derived:29 derived postfini
derived:30 derived postfini
derived:31 derived postfini
derived:32 derived postfini
derived:33 derived postfini
derived:34 derived postfini
derived:35 derived postfini
derived:36 derived postfini
derived:37 derived postfini
derived:38 derived postfini
derived:39 derived postfini
derived:40 derived postfini
derived:41 derived postfini
